src/dataflow/src/source/timestamp.rs

```diff
@@ -0,0 +1,259 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Types and methods for managing timestamp assignment and invention in sources
+
+use std::cell::RefCell;
+use std::collections::{HashMap, HashSet};
+use std::rc::Rc;
+
+use timely::progress::frontier::{Antichain, AntichainRef, MutableAntichain};
+
+use dataflow_types::MzOffset;
+use expr::{GlobalId, PartitionId};
+use repr::Timestamp;
+
+/// This struct holds per-source timestamp state in a way that can be shared across
+/// different source instances and allow different source instances to indicate
+/// how far they have read up to.
+///
+/// This type is almost never meant to be used directly, and you probably want to
+/// use `TimestampBindingRc` instead.
+pub struct TimestampBindingBox {
+    /// List of timestamp bindings per independent partition. This vector is sorted
+    /// by timestamp and offset and each `(time, offset)` entry indicates that offsets <=
+    /// `offset` should be assigned `time` as their timestamp. Consecutive entries form
+    /// an interval of offsets.
+    partitions: HashMap<PartitionId, Vec<(Timestamp, MzOffset)>>,
+    /// Indicates the lowest timestamp across all partitions that we retain bindings for.
+    /// This frontier can be held back by other entities holding the shared
+    /// `TimestampBindingRc`.
+    compaction_frontier: MutableAntichain<Timestamp>,
+}
+
+impl TimestampBindingBox {
+    fn new() -> Self {
+        Self {
+            partitions: HashMap::new(),
+            compaction_frontier: MutableAntichain::new(),
+        }
+    }
+
+    fn adjust_compaction_frontier(
+        &mut self,
+        remove: AntichainRef<Timestamp>,
+        add: AntichainRef<Timestamp>,
+    ) {
+        self.compaction_frontier
+            .update_iter(remove.iter().map(|t| (*t, -1)));
+        self.compaction_frontier
+            .update_iter(add.iter().map(|t| (*t, 1)));
+    }
+
+    fn compact(&mut self) {
+        let frontier = self.compaction_frontier.frontier();
+
+        // Don't compact up to the empty frontier as it would mean there were no
+        // timestamp bindings available
+        // TODO(rkhaitan): is there a more sensible approach here?
```

Comment:
An empty frontier indicates that the source will not be used again. If there is an empty frontier and someone does use the source again, a bug has happened somewhere.